Faster lightcurve __init__

omargamal8 opened the pull request:
This is a part of a solution to the issue no #164.
The commented function is used to check and issue a warning if the given bins are not of the same sizes.
But it takes very very longtime (~160s when used with 10^8 bins). By removing it, the run time decreased from 220s to 60s.
So I'd like to discuss an alternative way we can achieve the check with less cost.

![with_allclose2](https://cloud.githubusercontent.com/assets/23017651/24582641/4624cadc-1734-11e7-967f-424b0667a279.png)

![without_allclose](https://cloud.githubusercontent.com/assets/23017651/24582640/4602d508-1734-11e7-806e-49eb838f4d80.png)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/stingraysoftware/stingray/180)
<!-- Reviewable:end -->

matteobachetti commented:
@dhuppenkothen This is a problem. Light curve loading should be fast, and this operation is taking more than half the execution time for large datasets. 
I propose to disable the check by default, and use a switch (something like `robust=True`) when the user wants to be extra sure. We might let `simon` warn that for critical datasets one might want to use the switch. Or maybe, less favorite, enable the check by default but making the user know that for large dataset there is an alternative solution.

omargamal8 commented:
Or maybe we can separate it to be a separate member function, maybe named checkForEqualSizes.
This way it will be used only when called explicitly.

dhuppenkothen commented:
I quite like a combination of those, actually: pull the check out into its own method that people can use to check whenever they need to, disable by default and include a `SIMON` warning that for uneven bins many of the Fourier methods won't give useful results.

abigailStev commented:
The Travis and AppVeyor builds failed on this test:
```
    def test_irregular_time_warning(self):
        """
            Check if inputting an irregularly spaced time iterable throws out
            a warning.
            """
        times = [1, 2, 3, 5, 6]
        counts = [2, 2, 2, 2, 2]
        warn_str = ("SIMON says: Bin sizes in input time array aren't equal "
                    "throughout! This could cause problems with Fourier "
                    "transforms. Please make the input time evenly sampled.")
    
        with warnings.catch_warnings(record=True) as w:
            lc = Lightcurve(times, counts)
>           assert str(w[0].message) == warn_str
E           IndexError: list index out of range
stingray\tests\test_lightcurve.py:49: IndexError
```

omargamal8 commented:
@abigailStev  Yes, at that point it doesn't catch it because I commented out the function that checks for irregular time lapses without providing another solution.

Now with the new commit I made it will catch it if you override the new default parameter I added ("validate_dt"). Also I've changed the way the checking happens:
**Old Version:**
dt_array = np.diff(self.time)
if not (np.allclose(dt_array, np.repeat(self.dt, dt_array.shape[0]))):
simon("Bin sizes in input time array aren't equal throughout! "
            "This could cause problems with Fourier transforms. "
             "Please make the input time evenly sampled.") 
I suspect that this is slow because we use the array self.time (which is in this example 10^8 elements) to create from it the dt_array, and then we create another array having the same size by using the func np.repeat. I believe that this is causing too much overhead. 

Instead I made the check straight forward by going over the array once and checking for all the intervals:
**New version:**
for i in range (len(self.time)-1):
                   if( self.dt != (self.time[i+1] - self.time[i]) ):
                       simon("Bin sizes in input time array aren't equal throughout! "
                             "This could cause problems with Fourier transforms. "
                             "Please make the input time evenly sampled.")
                   break

I've tested the both versions using small array sizes and large ones too. The old version seems to have a slight advantage in small arrays but the new version out scales in the large arrays.

omargamal8 commented:
This is some analysis using Snake Viz. The number of bins in the lightcurve that was used in this example is 10^8

**Old Version**
![image](https://cloud.githubusercontent.com/assets/23017651/25819452/530b7a7c-342e-11e7-9417-702a4f7a24d4.png)



**New Version**
![image](https://cloud.githubusercontent.com/assets/23017651/25819363/1180942a-342e-11e7-9935-abfcc464b48a.png)

matteobachetti commented:
Hi @omargamal8, I suspect this could be a very good application of `numba`'s `jit` or another llvm compiler. That loop can be made into a very simple function that returns True or False, just iterating over the array, and based on the return value throw a `simon` warning. Can you make a try? 
We will then discuss if it's worth adding another dependency (numba), maybe optional.

omargamal8 commented:
@matteobachetti  Yes, I was thinking of creating multiple threads to perform the check faster. 
But first am trying to resolve a small problem am facing in comparing the numbers using the normal "==", it sometimes mess up.
I was comparing 0.03125 == 0.03125 and it gives out False.
If I use numpy.isclose to compare .. It consumes a lot of time.

matteobachetti commented:
Comparing floats with `==` is not a good idea. We should choose a tolerance and look for deviations from that tolerance. E.g. `thing_to_evaluate = np.abs(a - b) < tol`

omargamal8 commented:
@matteobachetti  you are right, I shouldn't compare them using ==, but am trying to avoid calling np.
I'll try implementing the same function np uses.
After that Ill use numba to optimize and see the results, Ill also use some distributed computing (either pyspark or ill use the raw multiprocessing). I'll provide the results.

dhuppenkothen commented:
@matteobachetti the `numpy` functions `numpy.isclose()` and `numpy.allclose()` do exactly what you're describing.

matteobachetti commented:
@dhuppenkothen I think there is some difference. Look here:
```
In [13]: a = np.random.normal(0, 1,  1000000)

In [14]: b = np.random.normal(a, 0.01)

In [15]: %time c = np.abs(a - b) < 0.1
CPU times: user 6.12 ms, sys: 8.33 ms, total: 14.4 ms
Wall time: 13.2 ms

In [16]: %time c = np.isclose(a, b, 0.1)
CPU times: user 16.8 ms, sys: 12.2 ms, total: 29 ms
Wall time: 27.9 ms

```

omargamal8 commented:
@dhuppenkothen  yes isclose and allclose both implements the exact functionality am looking for, but the problem is that we call it in each iteration throughout the 10^8 iteration. I might be wrong, but I think if we implement our own isclose it will be faster than calling a function in an external library. If it give us a slight advantage i.e 0.001 sec in each iteration, when u add that up to the 10^8 iteration, it will give a great advantage.

omargamal8 commented:
![image](https://cloud.githubusercontent.com/assets/23017651/25855070/72e85852-34d2-11e7-8f97-71656bd53545.png)

omargamal8 commented:
![isclosetest](https://cloud.githubusercontent.com/assets/23017651/25855625/01d79702-34d4-11e7-84ef-abd4db6c0edb.png)

matteobachetti commented:
@omargamal8 but using != is, as we told, not advisable with floats, there will be a lot of false positives. 
How long does the loop take using the absolute value and a threshold for the difference?

omargamal8 commented:
@matteobachetti Yes, I am just pointing out that time gap between the two. 
If I understood you correctly, You meant something like this right?
![screenshot from 2017-05-09 16-50-51](https://cloud.githubusercontent.com/assets/23017651/25857049/2a6c3020-34d8-11e7-8688-0c03e79e1fa7.png)

matteobachetti commented:
Yep! Looks good
